// stargate-mongoose/dist/driver/connection.js
"use strict";
// Copyright DataStax, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Connection = void 0;
const client_1 = require("../collections/client");
const collection_1 = require("./collection");
const connection_1 = __importDefault(require("mongoose/lib/connection"));
const connectionstate_1 = __importDefault(require("mongoose/lib/connectionstate"));
const utils_1 = require("../collections/utils");
class Connection extends connection_1.default {
    constructor(base) {
        super(base);
        this.debugType = 'StargateMongooseConnection';
        this.initialConnection = null;
    }
    _waitForClient() {
        return new Promise((resolve, reject) => {
            if ((this.readyState === connectionstate_1.default.connecting || this.readyState === connectionstate_1.default.disconnected) &&
                this._shouldBufferCommands()) {
                this._queue.push({ fn: resolve });
            }
            else if (this.readyState === connectionstate_1.default.disconnected && this.db == null) {
                reject(new Error('Connection is disconnected'));
            }
            else {
                resolve();
            }
        });
    }
    collection(name, options) {
        if (!(name in this.collections)) {
            this.collections[name] = new collection_1.Collection(name, this, options);
        }
        return super.collection(name, options);
    }
    async createCollection(name, options) {
        return (0, utils_1.executeOperation)(async () => {
            await this._waitForClient();
            const db = this.client.db();
            if (!this.client.httpClient.isAstra) {
                db.createDatabase();
            }
            return db.createCollection(name, options);
        });
    }
    async dropCollection(name) {
        return (0, utils_1.executeOperation)(async () => {
            await this._waitForClient();
            const db = this.client.db();
            return db.dropCollection(name);
        });
    }
    async dropDatabase() {
        return (0, utils_1.executeOperation)(async () => {
            await this._waitForClient();
            const db = this.client.db();
            return db.dropDatabase();
        });
    }
    async openUri(uri, options) {
        let _fireAndForget = false;
        if (options && '_fireAndForget' in options) {
            _fireAndForget = options._fireAndForget;
            delete options._fireAndForget;
        }
        // Set Mongoose-specific config options. Need to set
        // this in order to allow connection-level overrides for
        // these options.
        this.config = {
            autoCreate: options?.autoCreate,
            autoIndex: options?.autoIndex,
            sanitizeFilter: options?.sanitizeFilter,
            bufferCommands: options?.bufferCommands
        };
        for (const model of Object.values(this.models)) {
            // @ts-ignore
            model.init().catch(() => { });
        }
        this.initialConnection = this.createClient(uri, options)
            .then(() => this)
            .catch(err => {
            this.readyState = connectionstate_1.default.disconnected;
            throw err;
        });
        if (_fireAndForget) {
            return this;
        }
        await this.initialConnection;
        return this;
    }
    async createClient(uri, options) {
        this._connectionString = uri;
        this._closeCalled = false;
        this.readyState = connectionstate_1.default.connecting;
        const client = await client_1.Client.connect(uri, options);
        this.client = client;
        this.db = client.db();
        this.readyState = connectionstate_1.default.connected;
        this.onOpen();
        return this;
    }
    setClient(client) {
        this.client = client;
        this.db = client.db();
    }
    asPromise() {
        return this.initialConnection;
    }
    /**
     *
     * @returns Client
     */
    doClose(_force) {
        return this;
    }
}
exports.Connection = Connection;
