// stargate-mongoose/dist/collections/options.js
"use strict";
// Copyright DataStax, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.createCollectionOptionsKeys = exports.updateOneInternalOptionsKeys = exports.updateManyInternalOptionsKeys = exports.insertManyInternalOptionsKeys = exports.findOneAndUpdateInternalOptionsKeys = exports.findOneAndReplaceInternalOptionsKeys = exports.findInternalOptionsKeys = void 0;
class _FindOptionsInternal {
    constructor() {
        this.limit = undefined;
        this.skip = undefined;
        this.pagingState = undefined;
    }
}
exports.findInternalOptionsKeys = new Set(Object.keys(new _FindOptionsInternal));
/**
 * findOneAndReplaceOptions
 */
class _FindOneAndReplaceOptions {
    constructor() {
        this.upsert = undefined;
        this.returnDocument = undefined;
    }
}
exports.findOneAndReplaceInternalOptionsKeys = new Set(Object.keys(new _FindOneAndReplaceOptions));
/**
 * findOneAndUpdateOptions
 */
class _FindOneAndUpdateOptions {
    constructor() {
        this.upsert = undefined;
        this.returnDocument = undefined;
    }
}
exports.findOneAndUpdateInternalOptionsKeys = new Set(Object.keys(new _FindOneAndUpdateOptions));
/**
 * insertManyOptions
 */
class _InsertManyOptions {
    constructor() {
        this.ordered = undefined;
    }
}
exports.insertManyInternalOptionsKeys = new Set(Object.keys(new _InsertManyOptions));
/**
 * updateManyOptions
 */
class _UpdateManyOptions {
    constructor() {
        this.upsert = undefined;
    }
}
exports.updateManyInternalOptionsKeys = new Set(Object.keys(new _UpdateManyOptions));
/**
 * updateOneOptions
 */
class _UpdateOneOptions {
    constructor() {
        this.upsert = undefined;
    }
}
exports.updateOneInternalOptionsKeys = new Set(Object.keys(new _UpdateOneOptions));
/**
 * CreateCollectionOptions
 */
class _CreateCollectionOptions {
    constructor() {
        this.vector = undefined;
    }
}
exports.createCollectionOptionsKeys = new Set(Object.keys(new _CreateCollectionOptions));
